lib/class-wp-sourced-attributes-block-parser.php

```diff
@@ -0,0 +1,200 @@
+ *
+ * @package Gutenberg
+ */
+
+/**
+ * Class WP_Sourced_Attributes_Block_Parser
+ *
+ * Parses a document and constructs a list of parsed block objects
+ *
+ * @since 6.9.0
+ */
+class WP_Sourced_Attributes_Block_Parser extends WP_Block_Parser {
+
+	/**
+	 * Parses a document and returns a list of block structures
+	 *
+	 * When encountering an invalid parse will return a best-effort
+	 * parse. In contrast to the specification parser this does not
+	 * return an error on invalid inputs.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string                 $document Input document being parsed.
+	 * @param WP_Block_Type_Registry $registry Block type registry from which
+	 *                                         block attributes schema can be
+	 *                                         retrieved.
+	 * @param int|null               $post_id  Optional post ID.
+	 * @return WP_Block_Parser_Block[]
+	 */
+	function parse( $document, $registry = null, $post_id = null ) {
+		if ( is_null( $registry ) ) {
+			$registry = WP_Block_Type_Registry::get_instance();
+		}
+
+		$blocks = parent::parse( $document );
+
+		foreach ( $blocks as $i => $block ) {
+			$block_type = $registry->get_registered( $block['blockName'] );
+			if ( is_null( $block_type ) || ! isset( $block_type->attributes ) ) {
+				continue;
+			}
+
+			$sourced_attributes = $this->get_sourced_attributes(
+				$block,
+				$block_type->attributes,
+				$post_id
+			);
+
+			$blocks[ $i ]['attrs'] = array_merge( $block['attrs'], $sourced_attributes );
+		}
+
+		return $blocks;
+	}
+
+	/**
+	 * Returns an array of sourced attribute values for a block.
+	 *
+	 * @param WP_Block_Parser_Block $block             Parsed block object.
+	 * @param array                 $attributes_schema Attributes of registered
+	 *                                                 block type for block.
+	 * @param int|null              $post_id           Optional post ID.
+	 * @return array                                   Sourced attribute values.
+	 */
+	function get_sourced_attributes( $block, $attributes_schema, $post_id ) {
+		$attributes = array();
+
+		foreach ( $attributes_schema as $key => $attribute_schema ) {
+			if ( isset( $attribute_schema['source'] ) ) {
+				$attributes[ $key ] = $this->get_sourced_attribute(
+					$block,
+					$attribute_schema,
+					$post_id
+				);
+			}
+		}
+
+		return $attributes;
+	}
+
+	/**
+	 * Returns a sourced attribute value for a block, for attribute type which
+	 * sources from HTML.
+	 *
+	 * @param WP_Block_Parser_Block $block            Parsed block object.
+	 * @param array                 $attribute_schema Attribute schema for
+	 *                                                individual attribute to
+	 *                                                be parsed.
+	 * @return mixed                                  Sourced attribute value.
+	 */
+	function get_html_sourced_attribute( $block, $attribute_schema ) {
+		$document = new DOMDocument();
```

Comment:
It seems like this class has some [requirements](https://www.php.net/manual/en/dom.requirements.php). Can we confirm that the libxml package is currently a required one for WP Core?

Reply:
> It seems like this class has some [requirements](https://www.php.net/manual/en/dom.requirements.php). Can we confirm that the libxml package is currently a required one for WP Core?

From the page you link, it says "libxml is enabled by default".

We could still have some graceful fallback here for environments where it's explicitly disabled, although it would be unable to populate `$attributes`, yes.

Reply:
We may need to change the requires for WP core.